### PR TITLE
Change from private to protected for product variant

### LINKED ISSUE #2106: Change privates in product-variant service to protected
**Is your feature request related to a problem? Please describe.**
All the variables in the `product-variant` class are private, which means it can not be (easily) extended.

**Describe the solution you'd like**
Change them to protected (which is in line with the other services, like [product](https://github.com/medusajs/medusa/blob/master/packages/medusa/src/services/product.ts#L54-L63))

**Describe alternatives you've considered**
None

**Additional context**
Also would be a good opportunity to fix
Typo - [Incorrect js doc](https://github.com/medusajs/medusa/blob/master/packages/medusa/src/services/product-variant.ts#L143-L146) for SKU retrieval

### PR DESCRIPTION
**Pull request by troykelly**
- Change `private` to `protected` for class variables
- Correct JSDoc for Retrieve by SKU
- Add Retrieve by Barcode (self serving - can be removed if not needed)

Resolves #2106

### CODE REVIEW

File: packages/medusa/src/services/product-variant.ts [hunk 1]
```diff
@@ -63,19 +63,19 @@ class ProductVariantService extends BaseService {
   }) {
     super()
 
-    /** @private @const {EntityManager} */
+    /** @protected @const {EntityManager} */
```
Comment (srindom) [inline packages/medusa/src/services/product-variant.ts:66]: nit: we can actually remove these comments

**Review by srindom — COMMENTED** [review]
Thanks this is great! 

One thing that might be worth getting your view on too, is that we currently have a UPC, EAN and Barcode field on the ProductVariant db model. Wondering if it might be worth adding `retrieveByUpc` and `retrieveByEan` to not create confusion about what "barcode" refers to?

**Comment by troykelly** [comment]
> ... we currently have a UPC, EAN and Barcode field on the ProductVariant db model. Wondering if it might be worth adding `retrieveByUpc` and `retrieveByEan` to not create confusion about what "barcode" refers to?

I think you've highlighted a risk of confusion with what I have proposed - yes, it should be `byUpc` `byEan` etc

Fix as soon as I can

**Comment by troykelly** [comment]
@srindom Changes made. I think @olivermrbl wanted to avoid creating a function for each "type" of lookup that may be needed, but I do feel there's justification in making these three "quick" and accessible lookups.

**Comment by srindom** [comment]
I am fine with either - @olivermrbl wdyt?

@troykelly - can I ask you to get the branch up to date with develop :)

**Comment by olivermrbl** [comment]
@troykelly @srindom Also fine with including those methods. 

May I propose a slightly more sustainable approach used elsewhere:
```ts
async retrieve_(
    selector: Selector<ProductVariant>,
    config: FindConfig<ProductVariant> & PriceSelectionContext = {
      include_discount_prices: false,
    }
  ): Promise<ProductVariant> {
    const variantRepo = this.manager_.getCustomRepository(
      this.productVariantRepository_
    )

    const priceIndex = config.relations?.indexOf("prices") ?? -1
    if (priceIndex >= 0 && config.relations) {
      config.relations = [...config.relations]
      config.relations.splice(priceIndex, 1)
    }

    const query = this.buildQuery_(selector, config)
    const variant = await variantRepo.findOne(query)

    if (!variant) {
     const selectorConstraints = Object.entries(selector)
        .map(([key, value]) => `${key}: ${value}`)
        .join(", ")

      throw new MedusaError(
        MedusaError.Types.NOT_FOUND,
        `Product variant with ${selectorConstraints} was not found`
      )
    }

    return variant
  }
  
  async retrieveByEan(
    ean: string,
    config: FindProductVariantConfig = {}
  ): Promise<ProductVariant> {
    return await this.retrieve_({ ean }, config)
  }

  async retrieveByUpc(
    upc: string,
    config: FindProductVariantConfig = {}
  ): Promise<ProductVariant> {
    return await this.retrieve_({ upc }, config)
  }

  async retrieveBySKU(
    sku: string,
    config: FindProductVariantConfig = {}
  ): Promise<ProductVariant> {
    return await this.retrieve_({ sku }, config)
  }
  
  ...
```

**Comment by olivermrbl** [comment]
@troykelly Let me know if I should elaborate on the suggestion in my previous message :)

**Comment by troykelly** [comment]
Apologies @olivermrbl - been a little busy - will try and look at this in the afternoon (Sydney, AU)
I think your approach is far better than recreating the same code multiple times - thank you.

**Comment by srindom** [comment]
@troykelly - are you able to come back to this :)

**Comment by troykelly** [comment]
>  - are you able to come back to this :)
@srindom I will try to next week - apologies, it's been a hectic little while.
I should have some time over the Christmas break.

**Comment by riqwan** [comment]
Hey, thanks for the PR! Since v2 brought a lot of architectural and API changes on the backend, we will be closing this ticket since it no longer applies to our new setup.

**Pull request closed without merging**